conans/test/functional/devlayout/devlayout_test.py

```diff
@@ -0,0 +1,307 @@
+import os
+import platform
+import textwrap
+import unittest
+
+from nose.plugins.attrib import attr
+from parameterized import parameterized
+
+from conans.model.ref import ConanFileReference
+from conans.test.utils.tools import TestClient, TurboTestClient
+from conans.util.files import mkdir
+
+
+@attr("slow")
+class DevLayoutTest(unittest.TestCase):
+    conanfile = textwrap.dedent("""
+        from conans import ConanFile, CMake, CMakeLayout
+
+        class Pkg(ConanFile):
+            settings = "os", "compiler", "arch", "build_type"
+            options = {"shared": [True, False]}
+            default_options = {"shared": False}
+            generators = "cmake"
+            exports_sources = "src/*", "CMakeLists.txt"
+            generators = "cmake"
+            layout = "cmake"
+            
+            def build(self):
+                cmake = CMake(self) # Opt-in is defined having toolchain
+                cmake.configure()
+                cmake.build()
+
+            def package(self):
+                self.lyt.package()
+
+            def package_info(self):
+                self.lyt.package_info()
+                self.cpp_info.libs = ["hello"]
+            """)
+    cmake = textwrap.dedent("""
+        cmake_minimum_required(VERSION 2.8.12)
+        set(CMAKE_CXX_COMPILER_WORKS 1)
+        set(CMAKE_CXX_ABI_COMPILED 1)    
+        project(HelloWorldLib CXX)
+        
+        include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+        conan_basic_setup()
```

Comment:
It is important that tests use at least NO_OUPUT_DIRS in conan_basic_setup() if we are using layouts?

Reply:
It depends, the build-layout is not adjusting the output dirs, is **describing** where the artifacts will be. I'll add more tests to verify expected behavior.

Reply:
Added a test to illustrate how affects: `disable_output_dirs_test`